lib/TH/THGeneral.c

```diff
@@ -137,10 +159,17 @@ void* THRealloc(void *ptr, long size)
   if(size < 0)
     THError("$ Torch: invalid memory size -- maybe an overflow?");
 
-  ptr = realloc(ptr, size);
-  if(!ptr)
+  void *newptr = realloc(ptr, size);
+
+  if(!newptr && torchGCFunction) {
+    torchGCFunction(torchGCData);
+    newptr = realloc(ptr, size);
+  }
+
+  if(!newptr)
     THError("$ Torch: not enough memory: you tried to reallocate %dGB. Buy new RAM!", size/1073741824);
```

Comment:
tl;dr Both versions will leak if you catch the OOM within Lua, but lots of TH code isn't safe for this scenario.

@tkoeppe it depends on the context. Both `THAlloc` and `THRealloc` can call `THError` which in Lua calls `lua_error` and returns control to the Lua interpreter, which can catch the error in a `pcall`. So, if you want to avoid leaking memory, you need to make sure you remain in a consistent state if alloc never returns.

Note that this is not unique to `THRealloc`. Here's another example:
```C
void *p = THAlloc(n);
void *q = THAlloc(n); // OOM leaks p!
free(p);
free(q);
```

Correct code (for Lua) must assign all allocations to Lua-managed objects that will be cleaned up properly by the Lua GC. Here's a correct example from the code:

```C
tensor->size = THRealloc(self->size, sizeof(long)*nDimension);
```
`tensor` is a `THTensor*` sitting on the Lua stack, so `tensor->size` will be freed when `tensor` gets GC'd. If `THRealloc` never returns that's fine, the `tensor->size` still has the old pointer and will free it when it's GC'd.

Don't take this to mean that it's safe to catch OOM errors in a `pcall`. TH does *not* leave objects in a consistent state upon OOM. For example, consider generic/THTensor.c:667:

```C
self->size = THRealloc(self->size, sizeof(long)*nDimension);
self->stride = THRealloc(self->stride, sizeof(long)*nDimension);
self->nDimension = nDimension;
```
If `THRealloc` OOMs in one of the intermediate calls, `self` is broken. Calling `size()` will probably segfault. So this whole discussion is quite moot IMO.